libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -621,58 +621,121 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+		it = nextSignature;
 
-		// Only one function
-		if (next(begin) == end)
-			continue;
-
-		// Not an overridable function
-		if ((*it)->isConstructor())
+		if (baseFunctions.size() > 1)
 		{
-			for (begin++; begin != end; begin++)
-				solAssert((*begin)->isConstructor(), "All functions in range expected to be constructors!");
-			continue;
-		}
+			struct OverrideGraph {
+				std::map<FunctionDefinition const*, int> nodes;
+				std::map<int, FunctionDefinition const*> nodeInv;
+				std::map<int, std::set<int>> edges;
+				int numNodes = 2;
+				void addEdge(int a, int b)
+				{
+					edges[a].insert(b);
+					edges[b].insert(a);
+				}
+				int visit(FunctionDefinition const* function)
+				{
+					auto it = nodes.find(function);
+					if (it != nodes.end())
+						return it->second;
+					int currentNode = numNodes++;
+					nodes[function] = currentNode;
+					nodeInv[currentNode] = function;
+
+					if (!function->overrides())
+						addEdge(currentNode, 1);
+					else
+						for (auto const* baseFunction: function->annotation().baseFunctions)
+							addEdge(currentNode, visit(baseFunction));
 
-		// Function has been explicitly overridden
-		if (contains_if(
-			contractFuncs,
-			[&] (FunctionDefinition const* _f) {
-				return hasEqualNameAndParameters(*_f, **it);
-			}
-		))
-			continue;
+					return currentNode;
+				}
+			} overrideGraph;
+			for (auto const* baseFunction: baseFunctions)
+				overrideGraph.addEdge(0, overrideGraph.visit(baseFunction));
+
+			struct CutVertexFinder {
+				OverrideGraph const& graph;
+
+				std::vector<bool> visited = std::vector<bool>(graph.numNodes, false);
+				std::vector<int> depths = std::vector<int>(graph.numNodes, -1);
+				std::vector<int> low = std::vector<int>(graph.numNodes, -1);
+				std::vector<int> parent = std::vector<int>(graph.numNodes, -1);
+
+				std::set<int> cutVertices{};
+
+				void run(int u = 0, int depth = 0) {
+					visited.at(u) = true;
+					depths.at(u) = low.at(u) = depth;
+					int childCount = 0;
+					bool isCut = false;
+					for (int v: graph.edges.at(u))
+					{
+						if (!visited.at(v))
+						{
+							parent[v] = u;
+							run(v, depth + 1);
+							++childCount;
+							if (low[v] >= depths[u])
+								isCut = true;
+							low[u] = min(low[u], low[v]);
+						}
+						else if (v != parent[u])
+							low[u] = min(low[u], depths[v]);
+					}
+					if ((parent[u] != -1 && isCut) || (parent[u] == -1 && childCount > 1))
+						cutVertices.insert(u);
+				}
+			} cutVertexFinder{overrideGraph};
 
-		set<FunctionDefinition const*> ambiguousFunctions;
-		SecondarySourceLocation ssl;
+			cutVertexFinder.run();
 
-		for (;begin != end; begin++)
-		{
-			ambiguousFunctions.insert(*begin);
-			ssl.append("Definition here: ", (*begin)->location());
-		}
+			for (auto cutVertex: cutVertexFinder.cutVertices)
+			{
+				auto const* function = overrideGraph.nodeInv.at(cutVertex);
+				std::set<FunctionDefinition const*> toTraverse = function->annotation().baseFunctions;
```

Comment:
This seems like it is looking for functions above the cut vertex. Shouldn't it look down the inheritance graph?

Reply:
It looks above all cut vertices and removes everything from ``baseFunctions``. What's left afterwards is stuff that is *really* conflicting, so if more than one definition is left, we need to override.
This way it's easier to deal with multiple cut vertices (since I basically don't have to care).